src/spesh/optimize.c

```diff
@@ -2525,6 +2528,56 @@ static void try_eliminate_box_unbox_pair(MVMThreadContext *tc, MVMSpeshGraph *g,
         MVM_spesh_manipulate_delete_ins(tc, g, bb, ins);
     }
 }
+static void walk_set_looking_for_unbool(MVMThreadContext *tc, MVMSpeshGraph *g, MVMSpeshBB *bb,
+                                       MVMSpeshIns *box_ins, MVMSpeshIns *set_ins) {
+    MVMSpeshUseChainEntry *user_entry = MVM_spesh_get_facts(tc, g, set_ins->operands[0])->usage.users;
+    while (user_entry) {
+        MVMSpeshIns *user = user_entry->user;
+        if (user->info->opcode == MVM_OP_sp_runcfunc_i) {
+            MVMSpeshFacts *dispatch_facts = MVM_spesh_get_facts(tc, g, user->operands[1]);
+            if (dispatch_facts->flags & MVM_SPESH_FACT_KNOWN_VALUE
+                    && REPR(dispatch_facts->value.o)->ID == MVM_REPR_ID_MVMCFunction
+                    && ((MVMCFunction*)dispatch_facts->value.o)->body.func
+                        == MVM_disp_syscall_boolify_boxed_int_impl
+            ) {
+                try_eliminate_one_box_unbox(tc, g, bb, box_ins, user);
+            }
+        }
+        else if (user->info->opcode == MVM_OP_set ||
+                (user->info->opcode == MVM_SSA_PHI && user->info->num_operands == 2)) {
+            walk_set_looking_for_unbool(tc, g, bb, box_ins, user);
+        }
+        user_entry = user_entry->next;
+    }
+}
+static void try_eliminate_bool_unbool_pair(MVMThreadContext *tc, MVMSpeshGraph *g, MVMSpeshBB *bb,
+                                         MVMSpeshIns *ins, PostInlinePassState *pips) {
+    MVMSpeshUseChainEntry *user_entry = MVM_spesh_get_facts(tc, g, ins->operands[0])->usage.users;
+    while (user_entry) {
+        MVMSpeshIns *user = user_entry->user;
+        if (user->info->opcode == MVM_OP_sp_runcfunc_i) {
+            MVMSpeshFacts *dispatch_facts = MVM_spesh_get_facts(tc, g, user->operands[1]);
+            if (dispatch_facts->flags & MVM_SPESH_FACT_KNOWN_VALUE
+                    && REPR(dispatch_facts->value.o)->ID == MVM_REPR_ID_MVMCFunction
+                    && ((MVMCFunction*)dispatch_facts->value.o)->body.func
+                        == MVM_disp_syscall_boolify_boxed_int_impl) {
+                try_eliminate_one_box_unbox(tc, g, bb, ins, user);
+            }
+            else
+                return;
+        }
+        else if (user->info->opcode == MVM_OP_set ||
+                (user->info->opcode == MVM_SSA_PHI && user->info->num_operands == 2)) {
+            walk_set_looking_for_unbool(tc, g, bb, ins, user);
+        }
+        user_entry = user_entry->next;
+    }
+    if (MVM_spesh_usages_is_used(tc, g, ins->operands[0])) {
+    }
+    else {
+        MVM_spesh_manipulate_delete_ins(tc, g, bb, ins);
+    }
```

Comment:
Just `if (!MVM_spesh_usages_is_used(tc, g, ins->operands[0]))` to avoid the empty if branch?